e2e/test/nodepool/nodepool_test.go

```diff
@@ -0,0 +1,381 @@
+	expcapiv1alpha3 "sigs.k8s.io/cluster-api/exp/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/e2e/env"
+	"github.com/giantswarm/azure-operator/v4/e2e/setup"
+	key2 "github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+func Test_Nodepool(t *testing.T) {
+	err := nodepool.Test(context.Background())
+	if err != nil {
+		t.Fatalf("%#v", err)
+	}
+}
+
+type Config struct {
+	ClusterID  string
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+	Provider   *Provider
+	NodePoolID string
+}
+
+type Nodepool struct {
+	clusterID  string
+	ctrlClient client.Client
+	logger     micrologger.Logger
+	provider   *Provider
+	nodePoolID string
+}
+
+func New(config Config) (*Nodepool, error) {
+	if config.ClusterID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ClusterID must not be empty", config)
+	}
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.NodePoolID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.NodePoolID must not be empty", config)
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+
+	s := &Nodepool{
+		clusterID:  config.ClusterID,
+		ctrlClient: config.CtrlClient,
+		logger:     config.Logger,
+		nodePoolID: config.NodePoolID,
+		provider:   config.Provider,
+	}
+
+	return s, nil
+}
+
+const LabelVmSize = "node.kubernetes.io/instance-type"
+
+func (s *Nodepool) Test(ctx context.Context) error {
+	clusterID := s.clusterID
+
+	setupNodePoolID := s.nodePoolID
+	setupNodePoolVmSize := env.AzureVMSize()
+	setupNodePoolReplicas := 2
+
+	newNodepoolID := "t3st"
+	newNodePoolVmSize := "Standard_D3_v2"
+	newNodePoolReplicas := 3
+
+	err := s.CreateNodePool(ctx, newNodepoolID, int32(newNodePoolReplicas), newNodePoolVmSize)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.WaitForNodesReady(ctx, newNodePoolReplicas+setupNodePoolReplicas)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.assertRightNumberOfNodes(ctx, setupNodePoolID, newNodepoolID, setupNodePoolReplicas, newNodePoolReplicas, setupNodePoolVmSize, newNodePoolVmSize)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.provider.AddWorker(ctx, clusterID, newNodepoolID)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.WaitForNodesReady(ctx, newNodePoolReplicas+setupNodePoolReplicas+1)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.assertRightNumberOfNodes(ctx, setupNodePoolID, newNodepoolID, setupNodePoolReplicas, newNodePoolReplicas+1, setupNodePoolVmSize, newNodePoolVmSize)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = s.provider.ChangeVmSize(ctx, clusterID, newNodepoolID, "Standard_D5_v2")
```

Comment:
This changes the virtual machine size used in the specified node pool, which means changing the `AzureMachinePool` object. This tests that reconciliation works fine when changing that object.